plugins/node/instrumentation-fs/src/instrumentation.ts

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as api from '@opentelemetry/api';
+import { isTracingSuppressed, suppressTracing } from '@opentelemetry/core';
+import {
+  InstrumentationBase,
+  InstrumentationNodeModuleDefinition,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import { VERSION } from './version';
+import {
+  CALLBACK_FUNCTIONS,
+  PROMISE_FUNCTIONS,
+  SYNC_FUNCTIONS,
+} from './constants';
+import type * as fs from 'fs';
+import type {
+  FMember,
+  FPMember,
+  CreateHook,
+  EndHook,
+  FsInstrumentationConfig,
+} from './types';
+
+type FS = typeof fs;
+
+const supportsPromises = parseInt(process.versions.node.split('.')[0], 10) > 8;
+
+export default class FsInstrumentation extends InstrumentationBase<FS> {
+  constructor(config?: FsInstrumentationConfig) {
+    super('@opentelemetry/instrumentation-fs', VERSION, config);
+  }
+
+  init(): InstrumentationNodeModuleDefinition<FS>[] {
+    return [
+      new InstrumentationNodeModuleDefinition<FS>(
+        'fs',
+        ['*'],
+        (fs: FS) => {
+          this._diag.debug('Applying patch for fs');
+          for (const fName of SYNC_FUNCTIONS) {
+            if (isWrapped(fs[fName])) {
+              this._unwrap(fs, fName);
+            }
+            this._wrap(
+              fs,
+              fName,
+              <any>this._patchSyncFunction.bind(this, fName)
+            );
+          }
+          for (const fName of CALLBACK_FUNCTIONS) {
+            if (isWrapped(fs[fName])) {
+              this._unwrap(fs, fName);
+            }
+            this._wrap(
+              fs,
+              fName,
+              <any>this._patchCallbackFunction.bind(this, fName)
+            );
+          }
+          if (supportsPromises) {
+            for (const fName of PROMISE_FUNCTIONS) {
```

Comment:
since some might be not available i think you should check that they are defined before wrapping them, it might be possible that user rely on the function existing or not for their apps

Reply:
`shimmer` does that for us already and exits early. Do you think I should double-check it here before calling wrap?

> it might be possible that user rely on the function existing or not for their apps

The instrumentation doesn't change the API. I don't think I understood you correctly here, but if the user relies on `rm` function existing(added in v14) for example, but runs v12 we can't and should not do anything about it.